File: 443. String Compression/Solution.py
class Solution:
    def compress(self, chars: List[str]) -> int:
        n, pointer, editPointer = len(chars), 0, 0
        while pointer < n:
            start = pointer
            while pointer < n - 1 and chars[pointer + 1] == chars[pointer]: pointer += 1
            chars[editPointer] = chars[start]
            if pointer - start > 0:
                editPointer += 1
                count, divisor, flag = pointer - start + 1, 1000, False
                while count > 0:
                    if count // divisor > 0:
                        chars[editPointer] = str(count // divisor)
                        if divisor > 1 and count % divisor == 0: flag = True
                        editPointer += 1
                        count -= (count // divisor) * divisor
                    divisor //= 10
                if flag:
                    while divisor > 0:
                        chars[editPointer] = "0"
                        editPointer += 1
                        divisor //= 10
                editPointer -= 1
            pointer += 1
            editPointer += 1
        return editPointer

File: 443. String Compression/Solution.java
class Solution {
    public int compress(char[] chars) {
        int n = chars.length, pointer = 0, editPointer = 0;
        while (pointer < n) {
            int start = pointer;
            while (pointer < n - 1 && chars[pointer] == chars[pointer + 1]) pointer++;
            chars[editPointer] = chars[start];
            if (pointer - start > 0) {
                editPointer++;
                int total = pointer - start + 1, divisor = 1000;
                boolean flag = false;
                while (total > 0) {
                    if (total / divisor > 0) {
                        chars[editPointer] = (char) (total / divisor + '0');
                        editPointer++;
                        if (divisor > 1 && total % divisor == 0) flag = true;
                        total -= (total / divisor) * divisor;
                    }
                    divisor /= 10;
                }
                if (flag) {
                    while (divisor > 0) {
                        chars[editPointer] = '0';
                        editPointer++;
                        divisor /= 10;
                    }
                }
                editPointer--;
            }
            pointer++;
            editPointer++;
        }
        return editPointer;
    }
}